### PR TITLE
97 rest api for tickers

### PR DESCRIPTION
**Pull request by vstavetski**

### CODE REVIEW

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/common/CrossConversion.kt [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+class CrossConversion(
+        val tickerDaoService: TickerDaoService = AppContext.tickerDaoService,
+        val base: String,
+        val quote: String,
+        val exchange: String,
+        val windowDuration: Long,
+        val timestamp: Long
+) {
+
+    fun calculate(): ConversionResult {
```
Comment (hleb-albau) [inline rest-api/src/main/kotlin/fund/cyber/markets/rest/common/CrossConversion.kt]: nice code, just add javadoc here with steps

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/HistoHandler.kt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        val quote = params["tsym"]?.stringValue()
+        var exchange = params["e"]?.stringValue()
+        var tryConversion = params["tryConversion"]?.booleanValue()
+        var limit = params["limit"]?.intValue()
+        var timestamp = params["toTs"]?.longValue()
+
+        if (base == null || quote == null || base == quote) {
+            handleBadRequest("Bad parameters", httpExchange)
+        }
+        if (exchange == null) {
```
Comment (hleb-albau) [inline rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/HistoHandler.kt]: can be done int dedfinition line via  ':? "ALL"'
same for below items 
[example](https://github.com/cyberFund/cyber-search/blob/master/search-api/src/main/kotlin/fund/cyber/search/handler/SearchHandler.kt)

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/PriceHandler.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    override fun handleRequest(httpExchange: HttpServerExchange) {
+
+        val params = httpExchange.queryParameters
+        val base = params["fsym"]?.stringValue()
+        val quotes = params["tsyms"]?.stringValue()?.split(",")
+        var exchange = params["e"]?.stringValue()
+        var tryConversion = params["tryConversion"]?.booleanValue()
+
+        if (base == null || quotes == null) {
+            handleBadRequest("Bad parameters", httpExchange)
```
Comment (hleb-albau) [inline rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/PriceHandler.kt]: return?

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/PriceHandler.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            exchange = "ALL"
+        }
+        if (tryConversion == null) {
+            tryConversion = true
+        }
+
+        val timestamp = System.currentTimeMillis() / 60 / 1000 * 60 * 1000
+        val result = mutableMapOf<String, BigDecimal>()
+
+        for (quote in quotes!!) {
```
Comment (hleb-albau) [inline rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/PriceHandler.kt]: why !!  here?

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/AbstractHandler.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        val response = ErrorMessage(
+                "Error",
+                errorMessage
+        )
+        httpExchange.statusCode = 400
+        httpExchange.responseHeaders.put(Headers.CONTENT_TYPE, "application/json")
+        httpExchange.responseSender.send(jsonSerializer.writeValueAsString(response))
+    }
+
+    fun handleNoData(httpExchange: HttpServerExchange) {
```
Comment (hleb-albau) [inline rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/AbstractHandler.kt]: may be just 404?

File: build.gradle [hunk 6]
```diff
@@ -188,6 +189,13 @@ project(":connectors-gdax") {
 }
 
 project(":core") {
+
+    apply plugin: "kotlin-noarg"
+
+    noArg {
```
[the remark this answers is not in the document]
Comment (vstavetski) [reply build.gradle]: done

File: dao/src/main/kotlin/fund/cyber/markets/dao/service/TickerDaoService.kt [hunk 7]
```diff
@@ -2,20 +2,64 @@ package fund.cyber.markets.dao.service
...
+                "AND windowDuration=$windowDuration " +
+                "AND exchange='$exchange' " +
+                "AND timestampTo=$timestamp ")
+
+        val result = tickerMapper.map(resultSet)
+
+        return result.firstOrNull()
+    }
+
+    fun getTickers(base: String, quote: String, windowDuration: Long, exchange: String, timestamp: Long, limit: Int): List<Ticker> {
```
[the remark this answers is not in the document]
Comment (vstavetski) [reply dao/src/main/kotlin/fund/cyber/markets/dao/service/TickerDaoService.kt]: Yes, we can. I updated ticker service

File: dao/src/main/kotlin/fund/cyber/markets/dao/service/TickerDaoService.kt [hunk 8]
```diff
@@ -2,20 +2,64 @@ package fund.cyber.markets.dao.service
...
-    private val session: Session = cassandra.connect("markets").apply {
-        val manager = MappingManager(this)
-        tickerMapper = manager.mapper(Ticker::class.java)
+    fun insert(ticker: Ticker) {
+        tickerMapper.saveAsync(ticker)
     }
 
-    fun insert(ticker: Ticker) {
-        tickerMapper!!.saveAsync(ticker)
+    fun getTicker(base: String, quote: String, windowDuration: Long, exchange: String, timestamp: Long): Ticker? {
```
[the remark this answers is not in the document]
Comment (vstavetski) [reply dao/src/main/kotlin/fund/cyber/markets/dao/service/TickerDaoService.kt]: Updated ticker service

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/AbstractHandler.kt (hunk 5 above)
Comment (vstavetski) [reply rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/AbstractHandler.kt]: Changed status code to 404

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/PriceHandler.kt (hunk 3 above)
Comment (vstavetski) [reply rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/PriceHandler.kt]: Added return

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/PriceHandler.kt (hunk 4 above)
Comment (vstavetski) [reply rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/PriceHandler.kt]: refactor with constants values

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/HistoHandler.kt (hunk 2 above)
Comment (vstavetski) [reply rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/HistoHandler.kt]: done

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/common/CrossConversion.kt (hunk 1 above)
Comment (vstavetski) [reply rest-api/src/main/kotlin/fund/cyber/markets/rest/common/CrossConversion.kt]: done

File: dao/src/main/kotlin/fund/cyber/markets/dao/service/TickerDaoService.kt [hunk 9]
```diff
@@ -2,20 +2,64 @@ package fund.cyber.markets.dao.service
...
 
-    private var tickerMapper : Mapper<Ticker>? = null
+    private val session: Session = cassandra.connect("markets")
+    private val tickerMapper by lazy { MappingManager(session).mapper(Ticker::class.java) }
 
-    private val session: Session = cassandra.connect("markets").apply {
-        val manager = MappingManager(this)
-        tickerMapper = manager.mapper(Ticker::class.java)
+    fun insert(ticker: Ticker) {
+        tickerMapper.saveAsync(ticker)
```
[the remark this answers is not in the document]
Comment (vstavetski) [reply dao/src/main/kotlin/fund/cyber/markets/dao/service/TickerDaoService.kt]: added save confirmation

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/configuration/AppContext.kt [hunk 10]
```diff
@@ -0,0 +1,13 @@
+package fund.cyber.markets.rest.configuration
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import fund.cyber.markets.cassandra.CassandraService
+
+object AppContext {
+
+    val jsonSerializer = ObjectMapper()
+
```
Comment (YodaMike) [inline rest-api/src/main/kotlin/fund/cyber/markets/rest/configuration/AppContext.kt]: replace common ObjectMapper with the instance based variable due to performance issue in multithreaded applications

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/AbstractHandler.kt [hunk 11]
```diff
@@ -0,0 +1,38 @@
+package fund.cyber.markets.rest.handler
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import fund.cyber.markets.rest.configuration.AppContext
+import fund.cyber.markets.rest.model.ErrorMessage
+import io.undertow.server.HttpServerExchange
+import io.undertow.util.Headers
+
+open class AbstractHandler(
```
Comment (YodaMike) [inline rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/AbstractHandler.kt]: use abstract keyword

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/AbstractHandler.kt [hunk 12]
```diff
@@ -0,0 +1,37 @@
+package fund.cyber.markets.rest.handler
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import fund.cyber.markets.rest.model.ErrorMessage
+import io.undertow.server.HttpServerExchange
+import io.undertow.util.Headers
+
+abstract class AbstractHandler(
```
Comment (YodaMike) [inline rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/AbstractHandler.kt]: mark class as abstract

File: rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/PriceMultiFullHandler.kt [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+                                    ticker24h.maxPrice!!,
+                                    ticker24h.minPrice!!
+                            )
+                            quoteFullData.put(quote, priceData)
+                        }
+                    }
+                }
+                raw.put(base, quoteFullData)
+            }
+        }
```
Comment (YodaMike) [inline rest-api/src/main/kotlin/fund/cyber/markets/rest/handler/PriceMultiFullHandler.kt]: use comment
else {
// TODO: implement
}

**Pull request merged**